docs/architecture/adr-110-e2e-abci++.md

```diff
@@ -0,0 +1,267 @@
+```go
+func (app *Application) logRequest(req *abci.Request) {
+	s, err := GetABCIRequestString(req)
+	if err != nil {
+		panic(err)
+	}
+	app.logger.Debug(s)
+}
+```
+
+`GetABCIRequestString(req)` is a new method that receives a request and returns its string representation. The implementation and tests for this function and the opposite function `GetABCIRequestFromString(req)`
+that returns `abci.Request` from the string are provided in files `test/e2e/app/log.go` and `test/e2e/app/log_test.go`, respectively. To create a string representation of a request, we first marshal the request via `proto.Marshal()` method and then convert received bytes in the string using `base64.StdEncoding.EncodeToString()` method. The code of this method is below: 
+
+```go
+func GetABCIRequestString(req *abci.Request) (string, error) {
+	b, err := proto.Marshal(req)
+	if err != nil {
+		return "", err
+	}
+	reqStr := base64.StdEncoding.EncodeToString(b)
+	s := ABCI_REQ + reqStr + ABCI_REQ
+	return s, nil
+}
+```
+In addition, we surround the new string with `abci-req` constants so that we can find lines with ABCI++ request more easily.
+If in the future we want to log another ABCI++ request type, we just need to do the same thing: 
+create a corresponding `abci.Request` and log it via 
+`app.logRequest(r)`. 
+
+### 2) Parsing the logs
+We need a code that will take the logs from all nodes and collect the ABCI++ requests that were logged by the application. 
+
+<strong>Implementation</strong>
+
+This logic is implemented inside the `fetchABCIRequestsByNodeName()` function that resides in `test/e2e/tests/e2e_test.go` file. This function does three things:
+- Takes the output of all nodes in the testnet from the moment we launched the testnet until the function is called. It uses the `docker-compose logs` command. 
+- Parses the logs line by line and extracts the node name and the  `abci.Request`, if one exists. The node name is extracted manually and `abci.Request` is received by forwarding each line to the `app.GetABCIRequestFromString(req)` method.
+- Returns the map where the key is the node name, and the value is the list of all `abci.Request` logged on that node. 
+We can now use `[]*abci.Request` to store ABCI++ requests of any type, which is why we logged them in the previously described way. 
+
+ 
+
+### 3) ABCI++ grammar checker
+The idea here was to find a library that automatically verifies whether a specific execution respects the prescribed grammar. 
+
+<strong>Implementation</strong>
+
+We found the following library - https://github.com/goccmack/gogll. It generates a GLL or LR(1) parser and FSA-based lexer for any context-free grammar. What we needed to do is to rewrite ABCI++ grammar ([CometBFT's expected behaviour](../../spec/abci/abci%2B%2B_comet_expected_behavior.md#valid-method-call-sequences))
+using the synthax that the library understand. 
+The new grammar is below and can be found inside `test/e2e/pkg/grammar/abci_grammar.md` file.
+
+```abnf
```

Comment:
We've made small changes lately to the original grammar in the spec. Could you double-check this one is still aligned?

Reply:
Actually, you changed that now in case of a round with the `proposer` `ProcessProposal` is optional. From my understanding, you did this because of the process crashing scenario. Did you apply the same reasoning for the whole grammar?

Reply:
I think so